File: apps/desktop/src/lib/githubRateLimit.ts
/**
 * The rules for spending the GitHub quota, as pure functions.
 *
 * `stores/githubRateLimit.store.ts` holds what GitHub last said; this decides what to do about it.
 * Kept apart because every rule here is a judgement call worth testing on its own — which bucket a
 * URL is billed to, when to stop asking, how much to slow a poll down — and none of them needs a
 * store, a token or a React render to be checked.
 */

/**
 * Which of GitHub's independent allowances a request is billed to.
 *
 * Only the three the app can actually reach. `search` is the one worth naming separately: it is
 * 30 requests a *minute* against `core`'s 5000 an hour, so a handful of saved PR filters can exhaust
 * it while `core` is barely touched — and blocking diffs because the filters ran would be a bug.
 */
export type GithubRateResource = 'core' | 'search' | 'graphql'

/**
 * The bucket a URL will be billed to, decided before the request goes out.
 *
 * Known ahead of time because the throttle has to run *before* spending anything; GitHub's own
 * `x-ratelimit-resource` only arrives with the answer. The two agree by construction — every
 * `search/…` path is billed to `search`, `/graphql` to `graphql` — and the response's own value is
 * what the store records, so a disagreement corrects itself on the next call rather than persisting.
 */
export function rateLimitResourceForUrl(url: string): GithubRateResource {
  if (url.startsWith('https://api.github.com/graphql')) return 'graphql'
  if (url.startsWith('https://api.github.com/search/')) return 'search'
  return 'core'
}

/**
 * GitHub's own `x-ratelimit-resource`, narrowed to a bucket the app models — falling back to what
 * the URL implies for the ones it does not (`integration_manifest`, `code_search`, `dependency_
 * snapshots`…).
 *
 * Filing an unmodelled resource under the URL's bucket rather than inventing a key for it keeps the
 * throttle honest in the direction that matters: it will slow `core` down on the strength of a
 * neighbouring allowance, which costs a little speed, instead of tracking a bucket nothing ever
 * consults and letting the real one run to zero unnoticed.
 */
export function normalizeRateResource(reported: string, url: string): GithubRateResource {
  if (reported === 'core' || reported === 'search' || reported === 'graphql') return reported
  return rateLimitResourceForUrl(url)
}

/** What GitHub last said about one allowance, plus any cooldown the app imposed on itself. */
export interface GithubRateBucket {
  limit: number
  remaining: number
  /** Unix epoch **seconds**, as GitHub sends it. */
  reset: number
  /**
   * Epoch **milliseconds** until which no request should be sent, or `null` when free.
   *
   * Set either because the allowance is spent (wait for `reset`) or because GitHub answered a
   * secondary rate limit with a `Retry-After` — a separate mechanism from the hourly quota, and the
   * one GitHub explicitly documents as "wait this long before retrying".
   */
  blockedUntil: number | null
}

/**
 * Longest self-imposed cooldown, whatever the numbers say.
 *
 * `reset` is a timestamp from GitHub's clock compared against the local one, and the two can
 * disagree by more than a little on a machine whose time is off. An hour is the longest window any
 * bucket actually has, so a wait longer than that is a symptom of skew rather than of a real quota —
 * and the cost of capping it wrong is one request that comes back with the true numbers.
 */
export const MAX_COOLDOWN_MS = 60 * 60 * 1000

/** Milliseconds to wait before this bucket may be used again — `0` when it is free right now. */
export function cooldownMs(bucket: GithubRateBucket | undefined, now: number): number {
  if (!bucket?.blockedUntil) return 0
  return Math.min(Math.max(0, bucket.blockedUntil - now), MAX_COOLDOWN_MS)
}

/**
 * The cooldown a response justifies, or `null` to leave the bucket alone.
 *
 * Two distinct refusals, and conflating them would get one of them wrong:
 *
 * * A `Retry-After` is GitHub naming a duration — a *secondary* rate limit, triggered by making too
 *   many requests too quickly rather than by exhausting the hourly allowance. It is honoured
 *   verbatim, because ignoring it is what escalates a secondary limit into a longer one.
 * * A spent allowance (`remaining` at zero) is honoured until `reset`, because every request until
 *   then is a guaranteed 403 — and a 403 is what several callers turn into an empty list, so
 *   spending them means showing "you have no pull requests" over and over.
 *
 * A bucket with requests left imposes nothing: slowing down before the wall is the *poller's* job
 * (see {@link throttledInterval}), not a reason to refuse a merge the user just clicked.
 */
export function cooldownForResponse(
  res: { retryAfterSecs: number | null; rateLimit: { remaining: number; reset: number } | null },
  now: number
): number | null {
  if (res.retryAfterSecs != null) {
    return now + Math.min(res.retryAfterSecs * 1000, MAX_COOLDOWN_MS)
  }
  if (res.rateLimit?.remaining === 0) {
    return now + Math.min(Math.max(0, res.rateLimit.reset * 1000 - now), MAX_COOLDOWN_MS)
  }
  return null
}

/**
 * How much of a bucket is left, as a fraction, or `null` when GitHub has not been heard from.
 *
 * `null` is not "full": a bucket nobody has spent from yet and a bucket whose state is unknown lead
 * to the same behaviour here (poll normally), but only one of them is worth showing a user.
 */
export function remainingFraction(bucket: GithubRateBucket | undefined): number | null {
  if (!bucket || bucket.limit <= 0) return null
  return bucket.remaining / bucket.limit
}

/**
 * A poll interval slowed down in proportion to how little quota is left — `0` meaning "do not poll
 * at all", which is what SWR's `refreshInterval` reads as off.
 *
 * The app's polling is what spends the quota (a PR list every minute, an open PR's details every
 * thirty seconds, one search per saved filter), so it is what has to give way. Backing off in steps
 * rather than at a single threshold is deliberate: a cliff would let the app run at full speed until
 * it stopped dead, which is exactly the failure this exists to avoid. At a quarter left it polls
 * half as often, at a tenth four times less, and once the bucket is blocked it stops — by which
 * point the request would be refused anyway, so the only thing continuing would buy is a log line.
 */
export function throttledInterval(
  baseMs: number,
  bucket: GithubRateBucket | undefined,
  now: number
): number {
  if (cooldownMs(bucket, now) > 0) return 0
  const fraction = remainingFraction(bucket)
  if (fraction === null || fraction > 0.5) return baseMs
  if (fraction > 0.25) return baseMs * 2
  if (fraction > 0.1) return baseMs * 4
  return baseMs * 8
}

File: apps/desktop/src-tauri/src/services/github_rate_limit.rs
//! What GitHub says about how much of the quota is left, read off the `x-ratelimit-*` headers.
//!
//! The app polls: a pull-request list every minute, an open PR's details every thirty seconds, its
//! mergeability every twenty, one search per saved filter. None of that was ever measured against
//! the 5000 requests an hour a personal token gets — the headers were on every response and nothing
//! read them, so the first thing the user learned about exhausting the quota was a `403` that
//! several callers turn into an empty list. "You have no pull requests" is the worst possible way to
//! render "come back in nineteen minutes".
//!
//! Reading them costs nothing and gives the frontend two things it cannot derive: how close it is
//! (so polling can back off *before* the wall) and, on the far side of a secondary rate limit, the
//! `Retry-After` GitHub explicitly asks to be honoured.
//!
//! # The buckets are separate, and the difference matters
//!
//! `x-ratelimit-resource` names which allowance a response was billed to — `core` for ordinary REST,
//! `search` for `search/issues` (a much tighter 30/minute), `graphql` for the v4 endpoint. They are
//! independent: spending the search budget on saved PR filters must not stop the app from fetching a
//! diff. Anything that throttles on these numbers has to do it per resource, which is why the name
//! is carried through rather than collapsed here.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// GitHub's `x-ratelimit-*` headers for one response, parsed.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitHubRateLimit {
    /// Requests this bucket allows per window.
    pub limit: u32,
    /// Requests left in the current window. A `304` does not decrement it — that is the whole point
    /// of `github_etag_cache`.
    pub remaining: u32,
    /// Unix epoch seconds at which `remaining` returns to `limit`.
    pub reset: i64,
    /// Which allowance was billed: `core`, `search`, `graphql`… Defaults to `core` when GitHub omits
    /// it, which is what an older response looks like.
    pub resource: String,
}

/// Reads the quota headers, or `None` when the response carried none.
///
/// All three of limit/remaining/reset are required: two of them describe a state nothing can act on
/// (how close is "47 left" without knowing of what, or until when?), and a partial reading would be
/// worse than none — it would let a throttle make a decision from a number it half understands.
pub fn parse_rate_limit(
    limit: Option<&str>,
    remaining: Option<&str>,
    reset: Option<&str>,
    resource: Option<&str>,
) -> Option<GitHubRateLimit> {
    let limit = limit?.trim().parse::<u32>().ok()?;
    let remaining = remaining?.trim().parse::<u32>().ok()?;
    let reset = reset?.trim().parse::<i64>().ok()?;
    let resource = resource
        .map(str::trim)
        .filter(|r| !r.is_empty())
        .unwrap_or("core")
        .to_string();
    Some(GitHubRateLimit {
        limit,
        remaining,
        reset,
        resource,
    })
}

/// Parses `Retry-After` into a number of seconds to wait.
///
/// GitHub sends plain seconds on a secondary rate limit, but the header is also allowed to be an
/// HTTP date, so both are read — a date in the past (or a clock skewed the wrong way) yields `0`
/// rather than a negative wait, since "retry immediately" is the only sensible reading of it.
pub fn parse_retry_after(value: &str) -> Option<u64> {
    let value = value.trim();
    if value.is_empty() {
        return None;
    }
    if let Ok(secs) = value.parse::<u64>() {
        return Some(secs);
    }
    let at = DateTime::parse_from_rfc2822(value).ok()?;
    let delta = at.with_timezone(&Utc) - Utc::now();
    Some(delta.num_seconds().max(0) as u64)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reads_a_complete_set_of_headers() {
        let parsed = parse_rate_limit(Some("5000"), Some("4987"), Some("1789000000"), Some("core"))
            .expect("a reading");
        assert_eq!(parsed.limit, 5000);
        assert_eq!(parsed.remaining, 4987);
        assert_eq!(parsed.reset, 1789000000);
        assert_eq!(parsed.resource, "core");
    }

    #[test]
    fn keeps_the_resource_name_so_buckets_stay_separate() {
        // The tight one: `search` is 30/minute and independent of `core`'s 5000/hour.
        let parsed = parse_rate_limit(Some("30"), Some("2"), Some("1789000060"), Some("search"))
            .expect("a reading");
        assert_eq!(parsed.resource, "search");
    }

    #[test]
    fn defaults_the_resource_to_core_when_github_omits_it() {
        let parsed =
            parse_rate_limit(Some("5000"), Some("1"), Some("1789000000"), None).expect("a reading");
        assert_eq!(parsed.resource, "core");
        let blank = parse_rate_limit(Some("5000"), Some("1"), Some("1789000000"), Some("  "))
            .expect("a reading");
        assert_eq!(blank.resource, "core");
    }

    #[test]
    fn refuses_a_partial_reading() {
        // Half a quota is not a smaller quota — it is a number a throttle would misuse.
        assert_eq!(
            parse_rate_limit(Some("5000"), None, Some("1789000000"), None),
            None
        );
        assert_eq!(
            parse_rate_limit(None, Some("10"), Some("1789000000"), None),
            None
        );
        assert_eq!(parse_rate_limit(Some("5000"), Some("10"), None, None), None);
        assert_eq!(
            parse_rate_limit(Some("many"), Some("10"), Some("1789000000"), None),
            None
        );
    }

    #[test]
    fn reads_retry_after_as_seconds() {
        assert_eq!(parse_retry_after("60"), Some(60));
        assert_eq!(parse_retry_after(" 5 "), Some(5));
    }

    #[test]
    fn reads_retry_after_as_an_http_date_and_never_returns_a_negative_wait() {
        // Well in the past: the answer is "now", not a wait that would underflow.
        assert_eq!(parse_retry_after("Wed, 21 Oct 2015 07:28:00 GMT"), Some(0));
    }

    #[test]
    fn ignores_an_absent_or_unparseable_retry_after() {
        assert_eq!(parse_retry_after(""), None);
        assert_eq!(parse_retry_after("soon"), None);
    }
}
